src/webhook/test_namespace_validation.py

```diff
@@ -0,0 +1,122 @@
+import unittest
+import json
+
+from webhook import namespace_validation
+
+
+PRIVILEGED_NAMESPACES = (
+    "kube-admin",
+    "kube-foo",
+    "kube-",
+    "openshift",
+    "openshifter",
+    "openshift-foo",
+    "ops-health-monitoring",
+    "management-infra",
+    "default",
+    "logging",
+    "sre-app-check",
+    "redhat-user",
+    "redhat-",
+    "redhat-WOW",
+)
+
+NONPRIV_NAMESPACES = (
+    "kubeadmin",
+    "mykube-admin",
+    "Kube-Admin",
+    "open-shift",
+    "oopenshift",
+    "ops-health-monitoring-foo",
+    "the-ops-health-monitoring",
+    "management-infra1",
+    "mymanagement-infra",
+    "Default",
+    "default-user",
+    "adefault",
+    "logger",
+    "some-logging",
+    "sre-app-check$",
+    "redhatuser",
+)
+
+# None of these contain 'dedicated-admins', so will result in ALLOW unless
+# it's added.
+GROUP_LISTS = (
+    # These are tuples so they're immutable, forcing the test case to
+    # duplicate in order to change them. Otherwise we have potential
+    # collisions among test cases.
+    (),
+    ("cluster-admins",),
+    ("system:dedicated-admins", "sre-admins"),
+)
+
+
+def create_request(namespace, groups):
+    class FakeRequest(object):
+        json = {
+            "request": {
+                "uid": "testuser",
+                "userInfo": {
+                    "username": "me",
+                    "groups": groups,
+                },
+                "namespace": namespace,
+            }
+        }
+
+    return FakeRequest()
+
+
+class TestNamespaceValidation(unittest.TestCase):
+    def runtest(self, namespace, groups, expect):
+        # Make test failures easier to identify
+        failmsg = "expect={}, namespace={}, groups={}".format(
+            expect, namespace, groups)
+        request = create_request(namespace, groups)
+        response = namespace_validation.get_response(request)
+        response = json.loads(response)['response']
+        self.assertEqual(expect, response['allowed'], failmsg)
+        # On DENY, validate the status message
+        if not expect:
+            self.assertEqual(
+                "You cannot update the privileged namespace {}.".format(
+                    namespace),
+                response['status']['message'],
+                failmsg)
+
+    def test_deny(self):
+        # In order to get DENYs, we must have *both* a privileged namespace
+        # *and* the 'dedicated-admins' group.
+        for ns in PRIVILEGED_NAMESPACES:
+            for gl in GROUP_LISTS:
+                # Always include dedicated-admins
+                groups = gl + ('dedicated-admins',)
+                self.runtest(ns, groups, False)
+
+    def test_allow_group(self):
+        # If the group list doesn't contain 'dedicated-admins', always ALLOW,
+        # even if the namespace is privileged. (Really?)
```

Comment:
There's a problem here this brings up, tracked in https://issues.redhat.com/browse/OSD-3511
Beyond scope of this PR?

Reply:
I definitely feel that's for a different PR, yes.

(For that matter, I'm pretty unconcerned about the unrealistic-ness of some of the sample data we're testing with above. This is unit testing the code without making any assumptions about how the back end _really_ works.)